instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/aws.go

```diff
@@ -0,0 +1,102 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otelaws
+
+import (
+	"context"
+
+	v2Middleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
+	"github.com/aws/smithy-go/middleware"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+
+	"go.opentelemetry.io/contrib"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/semconv"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	tracerName = "go.opentelemetry.io/contrib/instrumentation/github.com/aws/aws-sdk-go-v2/otelaws"
+)
+
+type oTelMiddlewares struct {
+	tracer trace.Tracer
+}
+
+func (m oTelMiddlewares) initializeMiddleware(stack *middleware.Stack) error {
+	return stack.Initialize.Add(middleware.InitializeMiddlewareFunc("OtelSpanCreator", func(
+		ctx context.Context, in middleware.InitializeInput, next middleware.InitializeHandler) (
+		out middleware.InitializeOutput, metadata middleware.Metadata, err error) {
+
+		opts := []trace.SpanOption{
+			trace.WithSpanKind(trace.SpanKindClient),
+		}
+		ctx, span := m.tracer.Start(ctx, "AWS", opts...)
+		defer span.End()
+
+		out, metadata, err = next.HandleInitialize(ctx, in)
+		if err != nil {
+			span.RecordError(err)
+		}
+
+		return out, metadata, err
+	}),
+		middleware.Before)
+}
+
+func (m oTelMiddlewares) deserializeMiddleware(stack *middleware.Stack) error {
+	return stack.Deserialize.Add(middleware.DeserializeMiddlewareFunc("OtelSpanDecorator", func(
+		ctx context.Context, in middleware.DeserializeInput, next middleware.DeserializeHandler) (
+		out middleware.DeserializeOutput, metadata middleware.Metadata, err error) {
+		out, metadata, err = next.HandleDeserialize(ctx, in)
+		resp, ok := out.RawResponse.(*smithyhttp.Response)
+		if !ok {
+			// No raw response to wrap with.
+			return out, metadata, err
+		}
+
+		span := trace.SpanFromContext(ctx)
+		span.SetName(v2Middleware.GetServiceID(ctx))
```

Comment:
@speza set span name here.

Reply:
It seems a bit odd to now set the span name/operation to `AWS` to then mutate it later on? Would it be better to just use `v2Middleware.GetServiceID(ctx)` in `initializeMiddleware` but with `middleware.After` so that the span as at least been initialised with some metadata first?

Reply:
Agree that altering span name later is wield :) But `middleware.After` will loss the time of middlewares before OTelMiddleware.

Reply:
I think we're talking a tiny amount of time though right? I think the actions performed to the initializeMiddleware are all lightening fast and in memory? I guess theres a chance that they might not be one day?

Reply:
This is where I'm at with one for Datadog - https://github.com/DataDog/dd-trace-go/pull/871/files

Reply:
yeah, in previous PR I used `After`, but think about anyone can add his custom middleware and not sure how long it would take. Adding the OTel sdk middleware at the front can make sure to take monitoring responsibility :)

Reply:
I still think there is a problem here with setting the operation so late. Since this is the last middleware in the chain, there's many opportunities for a failure to happen before (e.g signing in FinalizeMiddleware). If that happens the trace will be pretty useless as it won't be clear as to what service was called?

Reply:
You could maybe register two Init middleware. One for Before and another for After? Start the trace in before, enrich it somewhat in the After

Reply:
> You could maybe register two Init middleware. One for Before and another for After? Start the trace in before, enrich it somewhat in the After. Then still use Deserialize for anything before/after the http request.

That is a good idea.

Reply:
@speza Please review the new push, thanks!

Reply:
One issue that may arise if the span name isn't available when `Start()` is called is that any samplers that depend on the name won't be able to function properly.  The two middleware approach sounds like a good way to handle it, though I'd suggest maybe just tracking the start timestamp in the simple `Before` middleware and starting a span with an explicit start time in `After` once the service ID is available to properly name the span as well.